### PR TITLE
React Aria Components API improvements

### PR DESCRIPTION
**Pull request by devongovett**
I've been making some examples with React Aria Components and Tailwind CSS, and came across a few things that were difficult to style. This PR includes some API changes meant to address them.

## Section headers

The `<Section>` component used within menus and listboxes previously accepted a `title` prop to render the header. This was one of the rare places where a component rendered multiple DOM elements. With something like Tailwind, this was a bit harder to style without using an [arbitrary variant](https://tailwindcss.com/docs/adding-custom-styles#arbitrary-variants) to target the `header` element inside the section.

This PR enables you to just render a `<Header>` element inside the `<Section>` yourself, removing the need for the `title` prop. This lets you easily style it.

```jsx
<Section>
  <Header className="...">Title</Header>
  <Item>Item</Item>
</Section>
```

For dynamic collections, you can use the `<Collection>` component to render the items.

```jsx
<Section>
  <Header>{section.name}</Header>
  <Collection items={section.children}>
    {item => <Item id={item.name}>{item.name}</Item>}
  </Collection>
</Section>
```

## Calendar header cells

`<CalendarGrid>` by default renders the entire month, and only lets you provide a function to render the cells. This means it's a bit challenging to style or customize the header cells (i.e. weekday names). Now you can optionally render a `<CalendarGridHeader>` and `<CalendarGridBody>` manually, which lets you render `<CalendarHeaderCell>` elements inside the header yourself and thus control the styling. This is optional - a function is still supported directly as children to CalendarGrid when this customization isn't needed.

```jsx
<CalendarGrid>
  <CalendarGridHeader>
    {day => <CalendarHeaderCell className="...">{day}</CalendarHeaderCell>}
  </CalendarGridHeader>
  <CalendarGridBody>
    {date => <CalendarCell date={date} />}
  </CalendarGridBody>
</CalendarGrid>
```

## Other

* Made separators actually accept a className when used inside a collection component like a menu. This works the same way as Header – there's a context that tells it that it is inside a collection, and it renders shallowly in that case. Then the collection component re-renders the real element later.
* Added `aria-expanded` to DatePicker buttons when the popover is open. I think this is correct anyway, but it is also useful for styling.
* Make `<SelectValue>` support `<Item>` elements with render prop children. Also pass the object value and textValue of the selected item to the custom render props of SelectValue itself.
* Apply the `data-hovered` and `data-focus-visible` states to more components, including `<Input>`, `<Group>`, `<DateInput>`, and `<SliderThumb>`.
* Apply the `data-pressed` state to buttons while an attached overlay is open. This was already true in some components and now it is true everywhere. Not totally sure if this is a good idea, or whether you'll want to have a different expanded state than pressed, but maybe it's a good default that you can override?

### CODE REVIEW

File: packages/react-aria-components/src/Group.tsx [hunk 1]
```diff
@@ -10,19 +10,60 @@
...
+  /**
+   * Whether an element within the group is focused, either via a mouse or keyboard.
+   * @selector :focus-within
+   */
+  isFocusWithin: boolean,
+  /**
+   * Whether an element within the group is keyboard focused.
+   * @selector [data-focus-visible]
+   */
+  isFocusVisible: boolean
```
Comment (devongovett) [inline packages/react-aria-components/src/Group.tsx:32]: Debated making this `isFocusVisibleWithin` but seemed like it was getting kinda long. Wondered about making the previous one `isFocused` for consistency even though it is within too, but we're using a native CSS selector for that so maybe confusing? 🤔

File: packages/react-aria-components/src/Header.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+function Header(props: HTMLAttributes<HTMLElement>, ref: ForwardedRef<HTMLElement>) {
+  [props, ref] = useContextProps(props, ref, HeaderContext);
+  let shallow = useShallowRender('header', props, ref);
+  if (shallow) {
+    return shallow;
+  }
+
+  return (
+    <header className="react-aria-Header" {...props} ref={ref}>
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/react-aria-components/src/Header.tsx:27]: Good catch!

File: packages/react-aria-components/docs/Calendar.mdx [hunk 3]
```diff
@@ -341,7 +374,39 @@ A [Button](Button.html) can be targeted with the `.react-aria-Button` CSS select
...
+
+<Calendar aria-label="Appointment date">
+  <header>
+    <Button slot="previous">◀</Button>
+    <Heading />
+    <Button slot="next">▶</Button>
+  </header>
+  <CalendarGrid>
+    <CalendarGridHeader>
+      {day => <CalendarHeaderCell style={{color: 'var(--blue)'}}>{day}</CalendarHeaderCell>}
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/react-aria-components/docs/Calendar.mdx:390]: heh, well orange doesn't have enough contrast with the background. I really struggled to find a color that would pass...

File: packages/react-aria-components/docs/ComboBox.mdx [hunk 4]
```diff
@@ -130,7 +130,7 @@ import {ComboBox, Label, Input, Button, Popover, ListBox, Item} from 'react-aria
     margin-top: 12px;
   }
 
-  .react-aria-Section header {
+  .react-aria-Header {
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/react-aria-components/docs/ComboBox.mdx:133]: hmm yeah. might think about this one for later. Right now most of the pages have the styling for everything at the top, so I think this feedback might apply beyond just headers...

File: packages/react-aria-components/docs/Calendar.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/react-aria-components/docs/Calendar.mdx:390]: argh, gotcha. I'm fine with it as is then

**Review by LFDanLu — APPROVED** [review]
LGTM

**Review by snowystinger — APPROVED** [review]

**Pull request merged**